spark-unified/src/main/scala/org/apache/spark/sql/delta/SparkTableConversion.scala

```diff
@@ -0,0 +1,125 @@
+
+import java.util.Optional
+
+import scala.collection.JavaConverters._
+
+import io.delta.kernel.spark.catalog.SparkTable
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.catalyst.plans.logical.{AppendData, LogicalPlan, OverwriteByExpression, OverwritePartitionsDynamic}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.delta.catalog.DeltaTableV2
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+
+/**
+ * Extractor for matching DataSourceV2Relation containing a SparkTable.
+ * Returns the DataSourceV2Relation and the SparkTable if matched.
+ */
+private object SparkTableRelation {
+  def unapply(relation: LogicalPlan): Option[(DataSourceV2Relation, SparkTable)] = {
+    relation match {
+      case dsv2 @ DataSourceV2Relation(sparkTable: SparkTable, _, _, _, _) =>
+        Some((dsv2, sparkTable))
+      case _ => None
+    }
+  }
+}
+
+/**
+ * An analyzer rule that converts SparkTable (Kernel-based V2 connector) to DeltaTableV2
+ * (legacy V1 connector) in DataSourceV2Relation nodes and V2 write commands.
+ *
+ * This conversion is necessary because SparkTable is a read-only implementation that
+ * lacks many features supported by DeltaTableV2. By converting SparkTable to DeltaTableV2
+ * early in the analysis phase, all existing Delta analysis rules (AppendDelta, OverwriteDelta,
+ * DeltaRelation, etc.) will work seamlessly.
+ *
+ * The conversion handles:
+ * - DataSourceV2Relation nodes (for reads and as children of other operators)
+ * - V2 write commands (AppendData, OverwriteByExpression, OverwritePartitionsDynamic)
+ *   where the table field is not a child and won't be transformed by resolveOperatorsDown
+ */
+class ConvertSparkTableToDeltaTableV2(session: SparkSession) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.resolveOperatorsDown {
+      // Handle DataSourceV2Relation nodes (for reads)
+      case SparkTableRelation(dsv2, sparkTable) =>
+        dsv2.copy(table = convertToDeltaTableV2(sparkTable))
+
+      // Handle AppendData (INSERT INTO)
+      case a: AppendData if isDeltaKernelTable(a.table) =>
+        val (dsv2, sparkTable) = extractSparkTable(a.table)
+        a.copy(table = dsv2.copy(table = convertToDeltaTableV2(sparkTable)))
+
+      // Handle OverwriteByExpression (INSERT OVERWRITE)
+      case o: OverwriteByExpression if isDeltaKernelTable(o.table) =>
+        val (dsv2, sparkTable) = extractSparkTable(o.table)
+        o.copy(table = dsv2.copy(table = convertToDeltaTableV2(sparkTable)))
+
+      // Handle OverwritePartitionsDynamic (dynamic partition overwrite)
+      case o: OverwritePartitionsDynamic if isDeltaKernelTable(o.table) =>
+        val (dsv2, sparkTable) = extractSparkTable(o.table)
+        o.copy(table = dsv2.copy(table = convertToDeltaTableV2(sparkTable)))
```

Comment:
Compared to `DeltaAnalysis`, the following logical plans are not handled:
* CreateTableLikeCommand (via EligibleCreateTableLikeCommand)
* CDCStatementBase
* TableChanges
* CloneTableStatement
* RestoreTableStatement
* DescribeDeltaHistory
* DeltaReorgTable
* ShowColumns
* DeltaMergeInto
* WriteToStream